ssl/ssl_lib.cc

```diff
@@ -1026,6 +1026,10 @@ static int ssl_read_impl(SSL *ssl) {
 }
 
 int SSL_read_ex(SSL *ssl, void *buf, size_t num, size_t *read_bytes) {
+  if (num == 0 && read_bytes != nullptr) {
```

Comment:
is the null pointer needed here?  Should we always set read_bytes to 0 and return 1 when num == 0?

Reply:
My goal here was to balance safety with compatibility. from my reading, OpenSSL won't return 1 in that case; it'll segfault. Admittedly, because of the amount of indirection via function pointers in OpenSSL's libssl, I'm not 100% confident in this reading of that code (see below).

So, if indeed my reading is correct and they segfault, I guess the expected behavior of passing null pointer here is "undefined" for OpenSSL. I thought it would be good for us to return an error in that case (as opposed to returning success) because the API's contract is for `*read_bytes` to be set to the number of bytes read on success. Since we can't do that with a null pointer, per the contract, returning success doesn't make sense in my opinion.

---

My reading of OpenSSL's relevant code:

- `SSL_read_ex` calls `ssl_read_internal`, which in turn [calls](https://github.com/openssl/openssl/blob/master/ssl/ssl_lib.c#L2328) the `ssl_read` implementation on `s->method`.
- I believe the implementation for our purposes here (i.e. not DTLS) [is](https://github.com/openssl/openssl/blob/master/ssl/ssl_local.h#L2294) `ssl3_write`, which in turn [calls](https://github.com/openssl/openssl/blob/master/ssl/s3_lib.c#L4510) an `ssl_write_bytes` implementation attached to `s->method`.
- Presumably, that implementation is `ssl3_write_bytes`, which never checks the `written` pointer before [dereferencing](https://github.com/openssl/openssl/blob/master/ssl/record/rec_layer_s3.c#L317) in the case when 0 bytes have been requested to be written (and 0 bytes have been written, i.e. `len == tot == 0` in the code).

Side  note -- somebody summed up the whole situation pretty well in [an internal comment](https://github.com/openssl/openssl/blob/master/ssl/ssl_local.h#L2118):

```
/*
 * This is for the SSLv3/TLSv1.0 differences in crypto/hash stuff It is a bit
 * of a mess of functions, but hell, think of it as an opaque structure :-)
 */
```